Add support for bootstrapping with an existing etcd cluster (#1364)
gardener/etcd-druid: docs/usage/bootstrapping-with-existing-cluster.md

# Bootstrapping with an Existing etcd Cluster

> This guide explains how to configure `spec.etcd.bootstrapWithExistingCluster` to make a new `Etcd` resource managed by `etcd-druid` join an already-running etcd cluster instead of starting as a standalone cluster.

Reviewer: ```suggestion
This guide explains how to configure `spec.etcd.bootstrapWithExistingCluster` to make a new `Etcd` resource managed by `etcd-druid` join an already-running etcd cluster (managed or not managed by etcd-druid) instead of starting as a standalone cluster.
```

> The conceptual background — terminology, lifecycle states, condition semantics, status shape, TLS trust model — is covered in [Bootstrap with an Existing etcd Cluster](../concepts/bootstrap-with-existing-cluster.md). This guide focuses on the **how**: prerequisites, step-by-step setup, and verification.

Reviewer: ```suggestion
The conceptual background — terminologies, lifecycle states, condition semantics, status shape, TLS trust model etc. is covered in [Bootstrap with an Existing etcd Cluster](../concepts/bootstrap-with-existing-cluster.md). This guide focuses on the **how**: prerequisites, step-by-step setup, and verification.
```

> [!NOTE]

> > This feature covers only the **join phase**: target members are added to the source cluster and synchronized. Removal of the original source members is tracked separately in [DEP-08](https://github.com/gardener/etcd-druid/pull/1355) and is **not implemented yet**. Until DEP-08 lands, the end state is a *combined* cluster of source + target members; if your goal is full migration, plan to remove source members manually for now.

Reviewer: ```suggestion
> This feature covers only the **join phase**: target members are added to the original source cluster and synchronized. Removal of the original source members is tracked separately in [DEP-08](https://github.com/gardener/etcd-druid/pull/1355) and is **not implemented yet**. Until DEP-08 lands, the end state is a *combined* cluster of source + target members; if your goal is full migration, plan to remove source members manually for now.
```

> ## Prerequisites

Reviewer: can you create a one more pre-requisite about version of etcd cluster.
1. If [UpgradeEtcdVersion feature-gate](https://github.com/gardener/etcd-druid/blob/master/docs/deployment/feature-gates.md#feature-gates-for-alpha-or-beta-features) is enabled then druid can only deploy the target etcd cluster of version 3.5.x then source cluster should also be version 3.5.x.
2. If [UpgradeEtcdVersion feature-gate](https://github.com/gardener/etcd-druid/blob/master/docs/deployment/feature-gates.md#feature-gates-for-alpha-or-beta-features) isn't enabled then druid can only deploy the target etcd cluster of version 3.4.x then source cluster should also be version 3.4.x

Before configuring `bootstrapWithExistingCluster` on a new `Etcd` resource, ensure:

> 1. **The source etcd cluster is healthy and reachable.** Every source member's peer URL must be dialable from the target's network namespace, and the source's client endpoint(s) must be reachable from the target's `etcd-backup-restore` sidecar.

Reviewer: ```suggestion
1. **All source etcd cluster members are healthy and reachable.** Every source member's peer URL must be dialable from the target's network namespace, and the source's client endpoint(s) must be reachable from the target's `etcd-backup-restore` sidecar.
```
it require all members to be in healthy state, not just quorum ... right ?

> 1. **The source etcd cluster is healthy and reachable.** Every source member's peer URL must be dialable from the target's network namespace, and the source's client endpoint(s) must be reachable from the target's `etcd-backup-restore` sidecar.

Reviewer 2: > Every source member's peer URL must be dialable from the target's network namespace

Is `Every` a requirement? from what I understand, it's only those extra additional advertise peer urls configured through `additionalAdvertisePeerUrls` need to be reachable from the target network right? because the original source domain specific urls need not be reachable by the target if I'm right? 🤔

> 2. **Source member names and peer URLs are known and stable.** You'll list them verbatim in the target spec; they must match what the source advertises in its own member list (`etcdctl member list`).

Reviewer: ```suggestion
2. **Source member names and peer URLs are known and stable.** You'll list them verbatim in the target etcd resource spec; they must match what the source advertises in its own member list (`etcdctl member list --endpoints=<source-client-endpoints>`).
```

```

Record each member's **`NAME`** and **`PEER ADDRS`** — they go into `bootstrapWithExistingCluster.members`. Record one or more **`CLIENT ADDRS`** for `bootstrapWithExistingCluster.clientEndpoints`.

> [!IMPORTANT]
> If source and target run in **different network domains** (separate Kubernetes clusters, on-prem ↔ cloud, etc.), the source must advertise peer URLs that are routable from the target. For etcd-druid managed sources, set `.spec.etcd.additionalAdvertisePeerURLs` on the source — see [Using Additional Advertise Peer URLs](using-additional-advertise-peer-urls.md). For externally managed sources, configure the source etcd's `--initial-advertise-peer-urls` directly. Use those externally-routable URLs in the target spec.

### 2. Author the target `Etcd` manifest

Add `bootstrapWithExistingCluster` to the target spec. A minimal 3-replica example:

```yaml
apiVersion: druid.gardener.cloud/v1alpha1
kind: Etcd
metadata:
  name: etcd-target
  namespace: target-ns
spec:
  replicas: 3
  etcd:
    image: <etcd-wrapper-image>
    bootstrapWithExistingCluster:
      members:
        - name: etcd-source-0
          peerUrls:

>             - https://etcd-source-0.etcd-source-peer.source-ns.svc:2380

Reviewer 2: should we be using examples of URLs which are externally available to route traffic from a different domain? because these internal headless dns urls are typically not reachable from outside, correct?

Reviewer 2: Since the purpose of this yaml is to serve as an example, can we also use multiple peer urls and use one out of those as externally available as mentioned in the comment above? This way we show diverse ways in which one can configure the creation of Etcd CR.

      clientEndpoints:
        - https://etcd-source-client.source-ns.svc:2379
    peerUrlTls:
      tlsCASecretRef:
        name: ca-etcd-peer-combined
        namespace: target-ns
      serverTLSSecretRef:
        name: etcd-target-server
        namespace: target-ns
    clientUrlTls:
      tlsCASecretRef:
        name: ca-etcd-client-combined
        namespace: target-ns
      serverTLSSecretRef:
        name: etcd-target-server-client
        namespace: target-ns
      clientTLSSecretRef:
        name: etcd-target-client
        namespace: target-ns
  backup: {}
  # ... other fields as needed
```

### 3. Apply the manifest

```bash
kubectl apply -f etcd-target.yaml
```

`etcd-druid` will:

- Append source member peer URLs to the target's `initial-cluster` ConfigMap entry.
- Append source `clientEndpoints` to the `--service-endpoints` flag of the target's `etcd-backup-restore` sidecar.
- Bring up target pods, which join the source cluster.

### 4. Watch the bootstrap progress

Track the dedicated condition:

```bash
kubectl -n target-ns get etcd etcd-target \
  -o jsonpath='{.status.conditions[?(@.type=="BootstrappedWithExistingCluster")]}{"\n"}'

> ```

Reviewer 2: Do you also want to show the example of how the field `.status.bootstrapWithExistingClusterMembers` will look like in case of both in progress and once it succeeds like how you've shown the status condition example?

While the join is in progress:

```json
{"type":"BootstrappedWithExistingCluster","status":"False","reason":"BootstrapInProgress","message":"Not all members have joined the cluster yet"}
```

Once it succeeds:

```json
{"type":"BootstrappedWithExistingCluster","status":"True","reason":"BootstrapSucceeded","message":"All members have successfully joined the existing cluster"}
```

The condition is **sticky**: once it flips to `True`, it stays `True` for the lifetime of the `Etcd` resource — pod restarts and `StatefulSet` rolls do not regress it.

### 5. Verify the combined cluster

> The `etcd-wrapper` container uses a [distroless](https://github.com/GoogleContainerTools/distroless) image and does not ship `etcdctl` or a shell, so `kubectl exec` cannot run `etcdctl` inside the pod. Use `kubectl port-forward` and run `etcdctl` from your local machine instead.

Reviewer 2: I think the recommended approach here should be to use the debug container as port-forwarding & pulling secrets out can be potentially dangerous if the user doesn't handle them well and expose the secrets. While the approach that's mentioned is generally acceptable, however i don't think we should be suggesting this in our official documentation. Wdyt?

Pull the target's client TLS material out of its `Secret`s so `etcdctl` can talk to the etcd client port over TLS:

```bash
kubectl -n target-ns get secret <client-ca-secret> \
  -o jsonpath='{.data.ca\.crt}' | base64 -d > /tmp/etcd-target-ca.crt
kubectl -n target-ns get secret <client-cert-secret> \
  -o jsonpath='{.data.tls\.crt}' | base64 -d > /tmp/etcd-target-client.crt
kubectl -n target-ns get secret <client-cert-secret> \
  -o jsonpath='{.data.tls\.key}' | base64 -d > /tmp/etcd-target-client.key
```

Replace `<client-ca-secret>` with `.spec.etcd.clientUrlTls.tlsCASecretRef.name` and `<client-cert-secret>` with `.spec.etcd.clientUrlTls.clientTLSSecretRef.name` from your manifest.

Forward the etcd client port from a target pod and list the members:

```bash
kubectl -n target-ns port-forward etcd-target-0 2379:2379 &

etcdctl --endpoints=https://localhost:2379 \
  --cacert=/tmp/etcd-target-ca.crt \
  --cert=/tmp/etcd-target-client.crt \
  --key=/tmp/etcd-target-client.key \
  member list -w table
```

> Expected: 6 members total — 3 source + 3 target — all `started`.

Reviewer: ```suggestion
Expected: 6 members total: 3 source members + 3 target members — all `started`.
```
can you also add one verification line with command that "you can verify that all members have joined the same cluster by checking the clusterID."

# Bootstrapping with an Existing etcd Cluster

This guide shows how to create a new etcd-druid managed `Etcd` resource that joins an already-running etcd cluster instead of starting as a standalone cluster.

> The source cluster may be managed by etcd-druid or externally managed. In both cases, the source member names, peer URLs, and client endpoints that you put into the target spec must already be valid from the target's network.

Reviewer: May you can add a new heading or one point for terminology and refer it to `terminology` of concepts doc because if I'm a user then I might directly come to usage doc, hence I can miss `terminologies` used in this doc.

For background and lifecycle semantics, see [Bootstrap with an Existing etcd Cluster](../concepts/bootstrap-with-existing-cluster.md).

## Prerequisites

Before creating the target `Etcd` resource, ensure:

> 1. **The source cluster is healthy.** All source members that you want the target to join with should be `started` in `etcdctl member list`.

Reviewer: > The source cluster is health. ... to join with should be `started` in `etcdctl member list`.

are you only referring to quorum or all members should be healthy because `started` meaning is not equal to `healthy`.

```

Record:

- each source member `NAME`,
- the source peer URLs that are reachable from the target pods,
- one or more source client endpoints reachable from the target backup-restore sidecar.

> [!IMPORTANT]
> The `PEER ADDRS` shown above are same-network examples. If source and target run in different network domains, replace them with peer URLs that the source cluster advertises and the target can dial.
>
> For an etcd-druid managed source, configure [`.spec.etcd.additionalAdvertisePeerURLs`](docs/usage/using-additional-advertise-peer-urls.md) on the source if extra reachable peer URLs are needed. For an externally managed source, prepare the advertised peer URLs outside etcd-druid. The target spec must use the source member names and peer URLs as advertised by the source member list.

## Step 2: Create the target `Etcd` manifest

Add `spec.etcd.bootstrapWithExistingCluster` when creating the target resource:

```yaml
apiVersion: druid.gardener.cloud/v1alpha1
kind: Etcd
metadata:
  name: etcd-target
  namespace: target-ns
spec:

>   replicas: 3

Reviewer: ```suggestion
  replicas: X // if source-cluster is running `Y` no. of replicas of etcd members, then total no. of etcd clusters member will become= X + Y
```
